module/VuFind/src/VuFind/Db/Table/ResourceTags.php

```diff
@@ -186,6 +186,84 @@ public function getResourcesForTag($tag, $userId, $listId = null)
         return $this->select($callback);
     }
 
+    /**
+     * Get lists associated with a particular tag.
+     *
+     * @param null|string|array $tag        Tag to match
+     * @param null|string|array $listId     ID of user owning favorite list
+     * @param null|string|array $userId     ID of list to retrieve
+     * @param bool              $publicOnly Whether to return only public lists
+     *
+     * @return \Laminas\Db\ResultSet\AbstractResultSet
+     */
+    public function getListsForTag(
+        $tag, $listId = null, $userId = null, $publicOnly = true
+    ) {
+        $tag = $tag ? (array)$tag : null;
+        $listid = $listId ? (array)$listId : null;
+        $userId = $userId ? (array)$userId : null;
+
+        $callback = function ($select) use (
+            $tag, $userId, $listId, $publicOnly
+        ) {
+            $select->columns(['list_id' => 'list_id', '*']);
+            $select->join(
+                ['t' => 'tags'],
+                'resource_tags.tag_id = t.id',
+                [
+                    'tag' =>
+                        $this->caseSensitive ? 'tag' : new Expression('lower(tag)')
+                ]
+            );
+            $select->join(
+                ['l' => 'user_list'],
+                'resource_tags.list_id = l.id',
+                []
+            );
+
+            // Discard tags assigned to a user resource.
+            $select->where->isNull('resource_id');
+
+            if ($listId) {
+                $select->where->and->in('resource_tags.list_id', $listId);
+            }
+            if ($userId) {
+                $select->where->and->in('resource_tags.user_id', $userId);
+            }
+            if ($publicOnly) {
+                $select->where->equalTo('public', 1);
+            }
+            if ($tag) {
+                if ($this->caseSensitive) {
+                    $select->where->and->in('t.tag', $tag);
+                } else {
+                    $lowerTags = array_map(
+                        function ($t) {
+                            return new Expression(
+                                'lower(?)', [$t], [Expression::TYPE_VALUE]
+                            );
+                        }, $tag
+                    );
+                    $select->where->and->in(
+                        new Expression('lower(t.tag)'), $lowerTags
+                    );
+                }
+            }
+            $select->group('resource_tags.list_id');
+
+            if ($tag) {
+                // Use AND operator for tags
+                $select->having->literal(
+                    'COUNT(resource_tags.list_id) = ?',
```

Comment:
Would it make more sense to `count(distinct(resource_tags.tag_id))` here? I'm concerned that counting the list_id might work now but might break in future if we extended the way we used the database (for example, if multiple users could add tags to a list).

Reply:
Good point, supporting tags also from non-list-owners (and filtering based on tag-creator) could be usable in the future. I tried `count(distinct(resource_tags.tag_id))` but it doesn't seem to work in a similar way. But could we achieve the same by adding an `where->in('resource_tags.user_id', $userId)` ? Btw: I accidentally used this in the previous version, although $userId should determine list-creator.... fixed now:
https://github.com/vufind-org/vufind/pull/1645/files#diff-40d3d2075f2211da8a41e54053a77d23R232

Reply:
Just to be sure I understand what you're trying to do here: the idea is that we want to AND together tag filters, rather than ORing them. When we say `$select->where->and->in('t.tag', $tag);`, we're bringing back all lists that have ANY of the requested tags... but when we `GROUP` by list_id, we want to apply a `HAVING` clause that will filter to only the lists that have ALL of the tags. If we can assume that each tag will be applied to each list only once, then we can filter on the groups that have a number of entries equal to the number of tags... but my concern is that this breaks down if there are scenarios where the same tag can be applied to a list multiple times by multiple users. To ensure that all requested tags are present, we would have to count distinct values, not just total number of rows. Some of the reading I've done suggests that this might work better if you add the count of distinct values as a selected value in the `columns()`, and then use the name of that value in the `HAVING` clause.

Does that help/make sense, or am I off base here?

Reply:
Yes, we want to AND together tags and OR ids. I revised the query to count by distinct tag id and use this in the HAVING clause. I also restricted the query to tags assigned by the list owner. Now the query should continue to work in the same way even if non-owners are allowed to tag lists in the future (which I think is a good thing to take into consideration).

About ANDing tags: this makes sense for our use-case, but should this be configurable? `ResourceTags::getListsForTag` could be passed the operator as a parameter and if OR is used, `GROUP BY`, `HAVING` and `COUNT(DISTINCT` could be dropped from the query.

Reply:
Thanks for the progress on this! It seems like it wouldn't hurt to add a "tagsOperator" setting to the .ini file and support it here by adding another parameter to the method. However, if this is more work than you want to spend time on, I think we can keep it as-is and work on that later as needed. I'll leave it up to you; just let me know your preference so I know whether or not to wait for more progress before doing a (hopefully) final round of testing.

Reply:
I added `tagsoperator`. We can leave it out if it makes things too complicated, but I think it could be useful.